src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,903 @@
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+CopyMakeBorder(CubeType& InImage, int top, 
+               int left, int bottom, int right)
+{
+  CubeType OutImage(InImage.n_rows + top + bottom, InImage.n_cols + left + right, InImage.n_slices);
+
+  for(size_t i = 0; i < InImage.n_slices; ++i)
+  {
+    OutImage.slice(i).submat(top, left, InImage.n_rows + top - 1, InImage.n_cols + left - 1)
+     = InImage.slice(i);
+    
+    for(size_t j = 0; j < right; ++j)
+    {
+      OutImage.slice(i).col(InImage.n_cols + left + j).subvec(top, InImage.n_rows + top - 1)
+      = InImage.slice(i).col(InImage.n_cols - j - 1);  
+    }
+
+    for(int j = 0; j < left; ++j)
+    {
+      OutImage.slice(i).col(j).subvec(top, InImage.n_rows + top - 1)
+      = InImage.slice(i).col(left - 1 - j);
+    }
+
+    for(int j = 0; j < top; j++)
+    {
+
+      OutImage.slice(i).row(j)
+      = OutImage.slice(i).row(2 * top - 1 - j);
+    }
+     
+    for(int j = 0; j < bottom; j++)
+    {
+      OutImage.slice(i).row(InImage.n_rows + top + j)
+      = OutImage.slice(i).row(InImage.n_rows + top - j - 1);
+    }
+
+  }
+  return OutImage;  
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+RGB2LUV(CubeType& InImage)
+{
+  //assert type is double or float.
+  double a, y0, maxi;
+  a = pow(29.0, 3) / 27.0;
+  y0 = 8.0 / a;
+  maxi = 1.0 / 270.0;
+
+  arma::vec table(1025);
+  for (size_t i = 0; i < 1025; ++i)
+  {
+    table(i) = i / 1024.0;
+    
+    if (table(i) > y0)
+      table(i) = 116 * pow(table(i), 1.0/3.0) - 16.0;
+    else
+      table(i) = table(i) * a;
+
+    table(i) = table(i) * maxi;
+  }
+
+  MatType rgb2xyz(3,3);
+  rgb2xyz(0,0) = 0.430574; rgb2xyz(0,1) = 0.430574; rgb2xyz(0,2) = 0.430574;
+  rgb2xyz(1,0) = 0.430574; rgb2xyz(1,1) = 0.430574; rgb2xyz(1,2) = 0.430574;
+  rgb2xyz(2,0) = 0.430574; rgb2xyz(2,1) = 0.430574; rgb2xyz(2,2) = 0.430574;
+
+  //see how to calculate this efficiently. numpy.dot does this.
+  CubeType xyz(InImage.n_rows, InImage.n_cols, rgb2xyz.n_cols);
+  for(size_t i = 0; i < InImage.n_rows; ++i)
+  {
+    for(size_t j = 0; j < InImage.n_cols; ++j)
+    {
+      for(size_t k = 0; k < rgb2xyz.n_cols; ++k)
+      {
+        double s = 0.0;
+        for(size_t l = 0; l < InImage.n_slices; ++l)
+          s += InImage(i, j, l) * rgb2xyz(l, k);
+        xyz(i, j, k) = s;
+      }
+    }
+  }
+
+  MatType nz(InImage.n_rows, InImage.n_cols);
+
+  nz = 1.0 / ( xyz.slice(0) + (15 * xyz.slice(1) ) + 
+       (3 * xyz.slice(2) + EPS));
+  
+  MatType L = arma::reshape(L, xyz.n_rows, xyz.n_cols);
+  
+  MatType U, V;
+  U = L % (13 * 4 * (xyz.slice(0) % nz) - 13 * 0.197833) + 88 * maxi;
+  V = L % (13 * 9 * (xyz.slice(1) % nz) - 13 * 0.468331) + 134 * maxi;
+
+  CubeType OutImage(InImage.n_rows, InImage.n_cols, InImage.n_slices);
+  OutImage.slice(0) = L;
+  OutImage.slice(1) = U;
+  OutImage.slice(2) = V;
+  //OutImage = arma::join_slices(L,U);
```

Comment:
Why not just put the LUV channels in the OutImage?

CubeType OutImage(InImage.n_rows, InImage.n_cols, InImage.n_slices);
OutImage.slice(0) = //?
OutImage.slice(1) = OutImage.slice(0) % (13 * 4 * (xyz.slice(0) % nz) - 13 * 0.197833) + 88 * maxi;
OutImage.slice(2) = OutImage.slice(0) % (13 * 9 * (xyz.slice(1) % nz) - 13 * 0.468331) + 134 * maxi;